docs/proposals/memory-management.md

```diff
@@ -0,0 +1,185 @@
+
+
+#### How is memory managed as of now?
+
+###### Requirements
+
+* Cgroups are used to manage memory on the node.
+* Kernel Out of Memory killer is expected to be enabled. [`/proc/<pid>/oom_score_adj`](https://www.kernel.org/doc/Documentation/filesystems/proc.txt) is used to manage priority for access to memory. The kernel assigns an `oom_score` to individual processes based on the memory usage of the process and a few [other policies](https://lwn.net/Articles/317814/). `oom_score_adj` is taken into account while calculating the oom_score. Higher values for `oom_score_adj` will increase the likelihood for being considered for OOM kills.
+
+###### Node level memory management
+
+Kubelet, the node agent, performs some amount of memory management. Specifically, it can
+
+1. Expose the memory capacity and usage of the node.
+
+2. Apply memory limits to user containers and individual processes on the node.
+
+3. Set up cgroups to track the memory consumption of itself and other system processes like docker daemon, kube-proxy, sshd, etc. The cgroups being setup by kubelet are,
+  a. `/docker-daemon`: The docker daemon is under this cgroup. This cgroup is limited to 70% of the total available memory on the node.
+  b. `/system`: OS processes, excepting init, are under this cgroup.
+  c. `/kubelet`: Kubelet is under this cgroup.
+  d. `/kube-proxy`: The kube-proxy is under this cgroup.
+
+4. Protect system processes from memory pressure on the node:
+   The `oom_score_adj` for kubelet, kube-proxy and other system processes is set to `-999`
+   The `oom_score_adj` for docker daemon is set to `-990`. Docker's memory usage has been unpredicatble in the past and docker daemons restarts are possible under system pressure.
+
+5. Manage Quality of service for user containers: Three classes of containers are currently supported:
+
+   a. `Guaranteed`: Memory is always guaranteed up to the request. These containers will not OOM unless they exceed their request, or system processes consume too much memory, or the kernel overhead is high and the memory in the node is being fully consumed. Note that `request` equals `limit` for this class of containers.
+
+   b. `Burstable`: Memory is guaranteed up until the usage exceeds user specified `request`. These containers do not get killed as soon as they exceed their request. Kernel memory overhead and system processes memory usage will impact memory availability for this class of containers. They are more likely to be killed by the kernel OOM killer once they exceed their request.
```

Comment:
thinking on this more, extending my scenario, container B continues to consume more memory on the node over time.  container A may never actually be able to restart.  it's only when container B is killed because of memory pressures on node that its possible any other container in the same class is no longer memory starved that something may restart.